services/device-registry-base/src/main/java/org/eclipse/hono/deviceregistry/service/device/AutoProvisioner.java

```diff
@@ -0,0 +1,386 @@
+    private Future<ProtonDelivery> sendAutoProvisioningEvent(
+            final String tenantId,
+            final String deviceId,
+            final String gatewayId,
+            final Span span) {
+
+        Objects.requireNonNull(tenantId);
+        Objects.requireNonNull(deviceId);
+
+        LOG.debug("sending auto-provisioning event for device [{}] created via gateway [{}] [tenant-id: {}]", deviceId, gatewayId, tenantId);
+
+        final Future<TenantResult<TenantObject>> tenantTracker = tenantInformationService.getTenant(tenantId, span);
+        final Future<DownstreamSender> senderTracker = downstreamSenderFactory.getOrCreateEventSender(tenantId);
+
+        return CompositeFuture.all(tenantTracker, senderTracker).compose(ok -> {
+            final Map<String, Object> props = new HashMap<>();
+            props.put(MessageHelper.APP_PROPERTY_QOS, QoS.AT_LEAST_ONCE.ordinal());
+            props.put(MessageHelper.APP_PROPERTY_DEVICE_ID, deviceId);
+            props.put(MessageHelper.APP_PROPERTY_TENANT_ID, tenantId);
+            props.put(MessageHelper.APP_PROPERTY_GATEWAY_ID, gatewayId);
+            props.put(MessageHelper.APP_PROPERTY_REGISTRATION_STATUS, EventConstants.RegistrationStatus.NEW.name());
+            props.put(MessageHelper.APP_PROPERTY_ORIG_ADAPTER, Constants.PROTOCOL_ADAPTER_TYPE_DEVICE_REGISTRY);
+            props.put(MessageHelper.APP_PROPERTY_ORIG_ADDRESS, EventConstants.EVENT_ENDPOINT);
+
+            final Message msg = MessageHelper.newMessage(
+                    ResourceIdentifier.from(EventConstants.EVENT_ENDPOINT, tenantId, deviceId),
+                    EventConstants.CONTENT_TYPE_EMPTY_NOTIFICATION,
+                    null,
+                    tenantTracker.result().getPayload(),
+                    props,
+                    null,
+                    false,
+                    false);
+
+            final DownstreamSender sender = senderTracker.result();
+            return sender.sendAndWaitForOutcome(msg, span.context())
+                    .onFailure(t -> LOG.info("error sending auto-provisioning event for device [{}] created via gateway [{}] [tenant-id: {}]", deviceId, gatewayId, tenantId));
+        });
+    }
+
+    /**
+     * Auto-provisions the edge device using the given device id and the given registration data.
+     *
+     * @param tenantId The id of the tenant for which the edge device should be provisioned.
+     * @param deviceId The id of the edge device which should be provisioned.
+     * @param gatewayId The id of the edge device's gateway.
+     * @param device The registration data for the device to be auto-provisioned.
+     * @param spanContext The tracing context to be used by this operation.
+     *
+     * @return A future indicating the outcome of the operation.
+     */
+    public Future<JsonObject> performAutoProvisioning(final String tenantId, final String deviceId,
+            final String gatewayId, final Device device, final SpanContext spanContext) {
```

Comment:
It is a _public_ method and FMPOV the arguments should be _null_ checked if applicable.